unread_msgs: Fix all unreads in muted stream being treated as muted.

prakhar1144 opened the pull request:
[CZO Discussion](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/register.3A.20.60unread_msgs.2Ementions.60/near/1645170)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details><summary>Before</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/c8d553c2-d654-4c65-9088-143f2c1ba92e">
</img>
</details> 

<details><summary>After</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/1c1b3e07-5d6d-49ff-b7eb-1ce748dfe9b1">
</img>
</details> 

The count is still not correct as we don't use `unread_msgs.count` directly;  PR to show count correctly: #26849

**Updated Docs**

<details><summary>Count</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/df91a579-837c-46f7-bd55-b1f5480400a7">
</img>
</details> 

<details><summary>Streams</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/35702ecc-36d5-47bd-87ff-b522dbffd246">
</img>
</details> 

<details><summary>mentions</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/2c764e1e-8e1d-4790-8ff0-88862f876357">
</img>
</details> 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented on zerver/lib/user_topics.py:
```diff
@@ -233,24 +233,26 @@ def mute_cond(row: RecipientTopicDict) -> ClauseElement:
         "recipient_id",
         "topic_name",
+        "visibility_policy",
     )
```
I think a single user doesn't have a very large number of user topic rows; so this shouldn't eat a lot of memory?

Or we should create three separate functions: `build_topic_mute_checker`, `build_topic_unmute_checker`, and `build_topic_follow_checker` ?

timabbott replied:
Yeah, I think this should be fine; and it's a set data structure, so performance should be fine too.

timabbott replied:
Thinking about this a bit more -- this still somewhat feels like the wrong API for how this is used, and since we only use it in one place, it might make sense to write it a bit more; with the same work, we could have this API be `get_prefetched_visibility_policy(recipient_id, topic) => visbiility_policy`, and call that function just once in `is_row_muted`; I don't love calling the same function 3 times with different parameters when we could just have the first line of `is_row_muted` be `visibility_policy = get_prefetched_visibility_policy_checker(recipient_id, topic)`

Since this is a loop through a very large number of messages, this sort of performance thing might matter.

prakhar1144 replied:
Makes sense.

Now, we have:
```
get_topic_visibility_policy = build_get_topic_visibility_policy(user_profile)

// within loop
visibility_policy = get_topic_visibility_policy(recipient_id, topic)
```
```
def build_get_topic_visibility_policy(
    user_profile: UserProfile,
) -> Callable[[int, str], int]:
    """Prefetch the visibility policies the user has configured for
    various topics.

    The prefetching helps to avoid the db queries later in the loop
    to determine the user's visibility policy for a topic.
    """
    rows = UserTopic.objects.filter(user_profile=user_profile).values(
        "recipient_id",
        "topic_name",
        "visibility_policy",
    )

    topic_to_visiblity_policy: Dict[Tuple[int, str], int] = defaultdict(int)
    for row in rows:
        recipient_id = row["recipient_id"]
        topic_name = row["topic_name"]
        visibility_policy = row["visibility_policy"]
        topic_to_visiblity_policy[(recipient_id, topic_name)] = visibility_policy

    def get_topic_visibility_policy(recipient_id: int, topic: str) -> int:
        return topic_to_visiblity_policy[(recipient_id, topic.lower())]

    return get_topic_visibility_policy
```

timabbott commented on zerver/lib/message.py:
```diff
@@ -1117,9 +1117,21 @@ def extract_unread_data_from_um_rows(
 
     def is_row_muted(stream_id: int, recipient_id: int, topic: str) -> bool:
-        if stream_id in muted_stream_ids:
+        stream_muted = stream_id in muted_stream_ids
```
Side note: We should probably change `get_muted_stream_ids` to return a `Set`; we're doing quadratic work here for no reason.

prakhar1144 replied:
Added a prep commit. -- 3rd commit

prakhar1144 commented:
Updates in the recent push:

1. Addressed the review comments
2. In `apply_unread_message_event()`, we had:

```
if ( stream_id not in state["muted_stream_ids"] and
     not topic_has_visibility_policy(user_profile, stream_id, topic, UserTopic.VisibilityPolicy.MUTED)
   ):
       state["unmuted_stream_msgs"].add(message_id)
```
Fixed it to consider a stream message as unmuted if it belongs to:
   * a not muted topic in a normal stream
   * an unmuted or followed topic in a muted stream

timabbott commented on zerver/lib/message.py:
```diff
@@ -1317,12 +1326,21 @@ def apply_unread_message_event(
             )
+            visibility_policy = user_topic.visibility_policy
+        except UserTopic.DoesNotExist:
+            visibility_policy = UserTopic.VisibilityPolicy.INHERIT
```
This bit should probably be extracted as a `get_visibility_policy(user_profile, stream_id, topic)` -- from an abstractions standpoint, we try pretty hard to avoid having lots of copies of queries that need to do `topic_name__iexact` -- we've had bad experiences in the past with mistakes in those, so try to keep that sort of query localized to functions inside the `topic` related modules:

```
$ git grep topic_name__iexact
zerver/lib/stream_subscription.py:                topic_name__iexact=topic_name,
zerver/lib/stream_topic.py:            topic_name__iexact=self.topic_name,
zerver/lib/stream_topic.py:            stream_id=self.stream_id, topic_name__iexact=self.topic_name
zerver/lib/user_topics.py:        topic_name__iexact=topic_name,
zerver/lib/user_topics.py:            user_profile=user_profile, stream_id=stream_id, topic_name__iexact=topic_name
zerver/lib/user_topics.py:        topic_name__iexact=topic_name,
zerver/lib/user_topics.py:        stream_id=stream_id, topic_name__iexact=topic_name
```

prakhar1144 replied:
Ok, abstracted it as 
```python
def get_topic_visibility_policy(
    user_profile: UserProfile,
    stream_id: int,
    topic_name: str,
) -> int:
    try:
        user_topic = UserTopic.objects.get(
            user_profile=user_profile, stream_id=stream_id, topic_name__iexact=topic_name
        )
        visibility_policy = user_topic.visibility_policy
    except UserTopic.DoesNotExist:
        visibility_policy = UserTopic.VisibilityPolicy.INHERIT
    return visibility_policy
```
in `user_topics.py`

We can use this function at a couple of other places in the "automatic follow/unmute PR". I will rebase + update that PR once this gets merged.

timabbott replied:
This is great, sounds good!

timabbott commented on zerver/lib/message.py:
```diff
@@ -1317,12 +1326,21 @@ def apply_unread_message_event(
+        # A stream message is unmuted if it belongs to:
+        # * a not muted topic in a normal stream
+        # * an unmuted or followed topic in a muted stream
+        if (not stream_muted and visibility_policy != UserTopic.VisibilityPolicy.MUTED) or (
```
This first part might be better as just `visibility_policy != UserTopic.VisibilityPolicy.MUTED` -- it's not actually important whether the stream is muted if the topic policy is `MUTED`, and that would I think make this a bit easier to read.

prakhar1144 replied:
When `visibility_policy = INHERIT`, using only `visibility_policy != UserTopic.VisibilityPolicy.MUTED` doesn't cover the case :

A message in a normal topic in a muted stream is not treated as unmuted message.

We can re-write it as
```
if not ((visibility_policy == MUTED) or (stream_muted and visibility_policy == INHERIT)):
   state["unmuted_stream_msgs"].add(message_id)
```

But I think, the existing logic is more readable ? We can change that if the later seems better to you.

timabbott replied:
Good catch, I like the logic you have, thanks for clarifying the bug my proposal would introduce :).

timabbott commented:
Posted a couple small comments on the details, otherwise lgtm!

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments.
* Updated the docs as [discussed on CZO](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/register.3A.20.60unread_msgs.2Ementions.60/near/1647324). -- updated images in the PR description.

timabbott commented:
Applied these changes to document as a normal API change, and marked to merge once CI passes:

``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index c8b0555df1..1edfeed023 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -20,6 +20,12 @@ format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 8.0
 
+**Feature level 213**
+
+* [`POST /register`](/api/register-queue): Fixed incorrect handling of
+  unmuted and followed topics in calculating the `mentions` and
+  `count` fields of the `unread_msgs` object.
+
 **Feature level 212**
 
 * [`GET /events`](/api/get-events), [`POST /register`](/api/register-queue),
diff --git a/version.py b/version.py
index 4ebbf00ba8..5bfed5f098 100644
--- a/version.py
+++ b/version.py
@@ -33,7 +33,7 @@ DESKTOP_WARNING_VERSION = "5.9.3"
 # Changes should be accompanied by documentation explaining what the
 # new level means in api_docs/changelog.md, as well as "**Changes**"
 # entries in the endpoint's documentation in `zulip.yaml`.
-API_FEATURE_LEVEL = 212
+API_FEATURE_LEVEL = 213
 
 # Bump the minor PROVISION_VERSION to indicate that folks should provision
 # only when going from an old version of the code to a newer version. Bump
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 48c6376040..757d880cdd 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -11854,8 +11854,8 @@ paths:
                               The total number of unread messages to display. This includes one-on-one and group
                               direct messages, as well as stream messages that are not [muted](/help/mute-a-topic).
 
-                              Zulip 8.0 fixed a bug where the "unmute topic" feature wasn't considered.
-                              ([#26848](https://github.com/zulip/zulip/pull/26848))
+                              **Changes**: Before Zulip 8.0 (feature level 213), the unmute and follow
+                              topic features were not handled correctly in calculating this field.
                           pms:
                             type: array
                             description: |
@@ -11954,8 +11954,8 @@ paths:
                               mentioned directly, and unread [non-muted](/help/mute-a-topic) messages
                               in which the user was mentioned through a wildcard.
 
-                              Zulip 8.0 fixed a bug where the "unmute topic" feature wasn't considered.
-                              ([#26848](https://github.com/zulip/zulip/pull/26848))
+                              **Changes**: Before Zulip 8.0 (feature level 213), the unmute and follow
+                              topic features were not handled correctly in calculating this field.
                             items:
                               type: integer
                           old_unreads_missing:
```

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -11897,8 +11899,9 @@ paths:
-                              messages of a single subscribed stream, including muted streams.
+                              An array of dictionaries where each dictionary contains details of all
+                              unread messages of a single subscribed stream. This includes muted streams
+                              and muted topics, even though those messages are excluded from count.
```
```suggestion
                              and muted topics, even though those messages are excluded from `count`.
```

timabbott replied:
That wasn't in time for the merge, so fixed in 9e3133e8200d408480cc9dce69cc0a418a93bf83.

gnprice reviewed (commented):
Thanks for these fixes! One nit below.
